src/Sylius/Bundle/ProductBundle/Doctrine/ORM/ProductAttributeValueRepository.php

```diff
@@ -27,4 +28,49 @@ public function findByJsonChoiceKey(string $choiceKey): array
             ->getResult()
         ;
     }
+
+    public function createByProductCodeAndLocaleQueryBuilder(
+        string $productCode,
+        string $localeCode,
+        ?string $fallbackLocaleCode = null,
+        ?string $defaultLocaleCode = null,
+    ): QueryBuilder {
+        $acceptableLocaleCodes = [$localeCode];
+
+        if (null !== $fallbackLocaleCode) {
+            $acceptableLocaleCodes[] = $fallbackLocaleCode;
+        }
+
+        if (null !== $defaultLocaleCode && array_count_values($acceptableLocaleCodes)[$localeCode] > 1) {
+            $acceptableLocaleCodes[] = $defaultLocaleCode;
+        }
+
+        $subQuery = $this->createQueryBuilder('s')
+            ->select('IDENTITY(s.attribute)')
```

Comment:
Could you justify usage of `IDENTITY` function here?

Reply:
The short justification is that I use this subquery in the `WHERE IN` clause, so I need IDs only.

I do it because I would like to filter out `product attribute values` with the appropriate locale. However, I fetch attributes with a fallback locale if there is no attribute value with the standard locale.

The thing I do in the main query is to fetch product attributes with the locale and fallback locale or null locale because it's another possible case. In other words: I fetch everything I can fetch based on the parameters the method receives. It's this part:
```
            ->andWhere(
                $queryBuilder->expr()->orX(
                    $queryBuilder->expr()->in('o.localeCode', $acceptableLocaleCodes),
                    $queryBuilder->expr()->isNull('o.localeCode')
                )
            )
```
Next, I don't want to have `product attribute values` with both locale and fallback locale. And here is the moment I do the `WHERE IN` subquery to have product attributes with fallback locale only if there is no product attribute with just the standard locale:
```
            ->andWhere(
                $queryBuilder->expr()->orX(
                    'o.localeCode = :locale',
                    $queryBuilder->expr()->notIn('IDENTITY(o.attribute)', $subQuery->getDQL())
                )
            )
```

I hope it's clear :smile: Do you think I should add the comment above this function that clarifies how the method exactly works?

Reply:
It's fine. Thx for clarification :)